DCOS-41419: Check available updates for installed UI packages

GeorgiSTodorov opened the pull request:
If a newer version of dcos-ui is available, inform the user.

Closes https://jira.mesosphere.com/browse/DCOS-41419

## Testing
Change `CHECK_DELAY` in `UpdateStream.ts` (on line 11) to be a smaller number, like 1000 or 2000;
Add `compare.subscribe(newVersion => console.log(newVersion));` at the end of `UpdateStream.ts`.
Open the UI and check the console for a message informing of a newer version of DCOS UI.

## Trade-offs
Some TODOs remain, updating them is in the scope of another ticket.

## Dependencies
None.

## Screenshots
Not applicable.

brandonc commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,81 @@
+  body: JSON.stringify({ includePackageVersions: true, packageName: "dcos-ui" })
+}).retry(4);
+
+export const compare = Observable.interval(CHECK_DELAY).switchMap(() =>
```
I don't think this will ever run until the app has been open for 24 hours. According to the code docs (http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-interval) the first emission is not sent immedately.

brandonc commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,81 @@
+    fetchedVersion,
+    (savedVersion, newVersionObject) => {
+      const newVersion = Object.keys(newVersionObject.response.results)[0];
+      if (newVersion > savedVersion) {
```
You may need a semantic version comparator, similar to the version sorting implementation that we use for sorting services by version. Reason being, the string "2.24.0" is greater than "2.100.0", but the semantic version is greater

brandonc replied:
Should you be comparing to the `window.DCOS_UI_VERSION` ?

brandonc commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,81 @@
+  );
+}
+
+setVersionInLocalStorage("2.24.0"); // TODO: Check if a version exists
```
This should contain the dismissed version numbers, no?

It might also be helpful to store the last time you checked for a new version, so you don't potentially trigger the notification every time the app loads.

GeorgiSTodorov replied:
I am not sure what you mean with dismissed version numbers. I think it should contain the last version that is installed. The current "2.24.0" is just a stub, the jira ticket says setting it is in the scope of another ticket.

nLight replied:
Local storage will contain versions that the UI has notified a user about and the version the user has _dismissed_. We don't do manually `setVersionInLocalStorage`. Current version comes from `window.DCOS_UI_VERSION` as @brandonc pointed it the prev comment.

GeorgiSTodorov replied:
@nLight Thank you, I understand now.

GeorgiSTodorov commented:
Thank you for the review, @brandonc .I addressed most of the feedback, I still need to write test and use `window.DCOS_UI_VERSION` as the current version.

GeorgiSTodorov commented:
I addressed the other feedback, @brandonc what kind of tests would you like to see?

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+);
+
+compare.subscribe(value => {
+  if (value) {
```
If we design the stream like `timer -> checkNewVersion -> filterDismissedVersions` then `subscribe` will fire only when there's a new version that has not yet been dismissed. So no `if (value)` needed.

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+    fetchedVersion,
+    (dismissedVersion, newVersionObject) => {
+      if (
+        new Date().getTime() <
```
Why do we need this conditional?

GeorgiSTodorov replied:
It's something @brandonc suggested and I liked the idea.

nLight replied:
So we don't re-check if a user refreshes the browser?

GeorgiSTodorov replied:
No, because I think otherwise the 24-hour timer would be sort of useless, I don't think anyone will keep the UI 24 hours without restarting/refreshing. Do you think that we should?

nLight replied:
Sentry suggests that some folks keep the UI open over weekends

nLight replied:
Regardless I'd suggest rewriting this check from being imperative to an appropriate rxjs operator like `filter`

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+  // When dismissing, save dismissed version in local storage
+  const currentVersion = window.DCOS_UI_VERSION;
+  console.log(
+    "A new version of DCOS UI is available. Upgrade " +
```
You can always use template strings 
```js
`A new version of DCOS UI is available. Upgrade ${currentVersion} to ${newVersion}. Dismiss.`
```

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+    dismissedVersion: getFromLocalStorage("dismissedVersion") || null,
+    lastTimeCheck: getFromLocalStorage("lastTimeCheck") || null
+  };
+  savedStates[DCOS_UI_VERSION][key] = value;
```
Why do you think it's beneficial to store dismissed version based of `DCOS_UI_VERSION`?

brandonc replied:
This is the global shadowing/confusing I was talking about ;)

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+import { request } from "@dcos/http-service";
+import Util from "#SRC/js/utils/Util";
+import { SAVED_STATE_KEY } from "#SRC/js/constants/UserSettings";
+import compareVersions from "compare-versions";
```
nitpick please put the `package.json` imports first and `#SRC` imports afterwards.

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+  )
+);
+
+compare.subscribe(value => {
```
I understand this is of debugging? Please don't subscribe to it right away in the production version.

GeorgiSTodorov replied:
When should we subscribe in production?

nLight replied:
That's a good question. Let me try to explain how I see this.

Ideally the whole update addition should be a plugin. Then I'd subscribe on the plugin bootstrap. I also can see us having "Disable updates" feature that would mean we never subscribe. 

Although that fantasy doesn't answer your question, I hope it hints why not subscribe right beside the declaration.

Working with what we have today, I'd say `pages/Index.js`? 

I would also like to consider wrapping an old-school plugin around the UI part. Like `plugins/banner` as a bare bone example. Leveraging the plugin we could also give an opportunity for a cluster admin to opt out from using the service entirely.

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+}
+
+setInLocalStorage("dismissedVersion", "2.24.2"); // stub
+const dismissedVersion = Observable.of(
```
We can use `startsWith` for default values http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-startWith

nLight replied:
I think by declaring it like that we effectively pin the version that was in the localStore on browser load and never re-check it. 

Why not making `getFromLocalStorage` observable with `Observable.create`?

nLight commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+
+export const compare = Observable.timer(0, CHECK_DELAY).switchMap(() =>
+  Observable.combineLatest(
+    dismissedVersion,
```
I wonder if `dismissedVersion` will be re-checked every time the timer fires

brandonc commented on src/js/core/UpdateStream.js:
```diff
@@ -0,0 +1,100 @@
+        new Date().getTime() <
+        getFromLocalStorage("lastTimeCheck") + CHECK_DELAY
+      ) {
+        return;
```
Is there a behavioral difference between return undefined and return false (below)?

GeorgiSTodorov replied:
Good point, there shouldn't be.

nLight commented:
I'm super sorry not bringing this up before but would you please rewrite the stream in Typescript. Thanks! 🙇

GeorgiSTodorov commented:
Started working on tests and did typescript migration, although I have a small problem with `UserSettingsStore.getKey` and `.setKey` not getting recognized by typescript (but still working) that I will look into tomorrow.

mperrotti commented:
Could you also implement `/api/v1/update/<version>` and `/api/v1/reset`?

mperrotti commented:
Nevermind that last comment. I think just linking to `/catalog/packages/{packageName}/deploy?version={newVersion}` will do what I need 👍

nLight commented on src/js/pages/Index.js:
```diff
@@ -72,6 +77,9 @@ var Index = React.createClass({
     ConfigStore.fetchCCID();
     ConfigStore.addChangeListener(EventTypes.CONFIG_ERROR, this.onConfigError);
+    compare.subscribe(values => {
+      showNotification(getVersionFromVersionObject(values[1]));
```
shall we `map` filtered result so that we could 
```js
compare.subscribe(newVersion => showNotification(newVersion))
```
?

GeorgiSTodorov replied:
Good point, done.

nLight commented on src/js/pages/Index.js:
```diff
@@ -20,6 +20,11 @@ import SidebarStore from "../stores/SidebarStore";
+import {
+  compare,
+  getVersionFromVersionObject,
+  showNotification
```
This function should live somewhere else. Not in the `UpdateStream`.

GeorgiSTodorov replied:
You are right, Mike and I even talked about this yesterday.

GeorgiSTodorov replied:
I will leave it there for now, until @mperrotti and I decide the best place to have it.

GeorgiSTodorov commented:
System tests in `test-universe.js` are broken, looking into them.

GeorgiSTodorov commented:
Let's see now.

DanielMSchmidt commented on src/js/constants/UserSettings.d.ts:
```diff
@@ -0,0 +1 @@
+export var SAVED_STATE_KEY: "savedStates";
```
🎨: could be const
```suggestion
export const SAVED_STATE_KEY: "savedStates";
```

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+import Util from "#SRC/js/utils/Util";
+import { SAVED_STATE_KEY } from "#SRC/js/constants/UserSettings";
+
+export const UpdateStreamType: symbol = Symbol("UpdateStreamType");
```
Nice use of symbols here 👍

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+import { SAVED_STATE_KEY } from "#SRC/js/constants/UserSettings";
+
+export const UpdateStreamType: symbol = Symbol("UpdateStreamType");
+const CHECK_DELAY: number = 24 * 60 * 60 * 1000; // Once every 24 hours
```
```suggestion
const CHECK_DELAY = 24 * 60 * 60 * 1000; // Once every 24 hours
```

DanielMSchmidt replied:
Not necessary, but it would be removing unnecessary information, making it more concise

GeorgiSTodorov replied:
Then we have to remove the type of all other `const`s like `DCOS_UI_VERSION_SETTING` as well.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+
+interface DCOSUserSettings {
+  dcosUIVersion: DCOSUIVersion;
+  [key: string]: DCOSUIVersion;
```
What is this `[key: string]` part about?

GeorgiSTodorov replied:
I searched for the error and I thought it was complaining about the key.

DanielMSchmidt replied:
The problem is that you are doing `savedStates[DCOS_UI_VERSION_SETTING] =  ...`. As some mutating method could have changed `DCOS_UI_VERSION_SETTING` typescript tries to tell you that there could be every key accessed. If you would change the line to `savedStates.dcosUIVersion` there would not be an error

GeorgiSTodorov replied:
Yes, but I wanted to make use of the constant.

GeorgiSTodorov replied:
It makes it easier to change `"dcosUIVersion"`.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+}
+
+function getVersionFromVersionObject(versionObject: any): string {
+  return Object.keys(versionObject.response.results)[0];
```
I think we have TS misconfigured, this should scream so much at you for accessing unknowns. What should this return if the versionsObject does not have a response or the response has no results or the results is an empty array? Some of those will return no string but a null, others will throw an error, which means that the type said to be returned here is wrong.

GeorgiSTodorov replied:
I added `interface VersionObject`, but then I got
![screenshot from 2018-11-20 14-47-18](https://user-images.githubusercontent.com/40791275/48774585-539f3f00-ecd3-11e8-9ddc-7d2f911fa54c.png)
.
And when I changed `RequestResponse`,
![screenshot from 2018-11-20 14-50-05](https://user-images.githubusercontent.com/40791275/48774720-aa0c7d80-ecd3-11e8-8a37-20e5ca740ea6.png)
. Observables with types are weird.

DanielMSchmidt replied:
So let me summarize what typescript tells you:

You told typescript that the request always returns an object on which the response key is exactly an empty object (by using `Observable<RequestResponse<{}>>`) and now you try to access on this empty object a result key, which is not there for the empty object, therefore the error.

If you set the `Observable<RequestResponse<{}>>` to something closer to the truth, maybe `Observable<RequestResponse<CosmosListResponse>>` and you give the `CosmosListResponse` a key attribute that is structured the same way as the HTTP response is then you should be better of. Keep in mind that if you don't know if a key is really always present you should add a ? `key?: number` or a null as union type `key: number | null`, to ensure that the type assertions are always true. After that, you need to alter the code in the map to be clearer with which types are always there and which ones are not.

If you find it confusing to work with typescript in observables I can recommend to use full functions with type signatures for every map, that way you can be sure of which type the outcome is.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+  return Object.keys(versionObject.response.results)[0];
+}
+
+function getFromLocalStorage(key: string): any {
```
Do we really have no idea what is returned here?

GeorgiSTodorov replied:
We do, but if I write `string | number` for example, then the `localStorageDismissedVersion` definition complains that
```
[ts]
Type 'BehaviorSubject<string | number>' is not assignable to type 'BehaviorSubject<string>'.
  Types of property 'value' are incompatible.
    Type 'string | number' is not assignable to type 'string'.
      Type 'number' is not assignable to type 'string'. [2322]
```
.

DanielMSchmidt replied:
Because the BehaviorSubject where you use this function only expects strings and no numbers. Either have this function return string or have the BehaviourSubject handle string | number.

In general, using the fewest amount, but the most specific types in a union type is a good rule of thumb

GeorgiSTodorov replied:
I don't like the solution of making both `BehaviorSubject`s `string | number`, but I don't see any other way, so I guess it's okay.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+    : 0
+);
+
+setInLocalStorage(DISMISSED_VERSION, "2.24.2"); // stub
```
Stub seems to be left over, I guess this is meant to be removed

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+> = new BehaviorSubject(
+  getFromLocalStorage(DISMISSED_VERSION)
+    ? getFromLocalStorage(DISMISSED_VERSION)
+    : "0"
```
Shouldn't this be sth like null instead? A string zero seems to be strange to identify as much as "nothing" (from my understanding)

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+
+export function showNotification(newVersion: string): void {
+  // TODO: Move inside UpdateServiceBanner
+  alert(newVersion);
```
This alert seems to be wrong here

DanielMSchmidt replied:
So we can't really merge this PR until mikes PR is done? We should note is as a dependency

GeorgiSTodorov replied:
It's the other way around. Mike uses this PR to make his stuff work.

GeorgiSTodorov replied:
It can still be reviewed separately, it just won't have a visible impact in the UI.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+}
+
+export function showNotification(newVersion: string): void {
+  // TODO: Move inside UpdateServiceBanner
```
❌ Isn't this part of this issue?

DanielMSchmidt replied:
Okay, gotcha, having the alert there and potentially merging it to master makes me worry though

GeorgiSTodorov replied:
It used to be `console.log`, but typescript doesn't allow those.

DanielMSchmidt replied:
Typescript or tslint?

DanielMSchmidt replied:
You can do a `// tslint:disable-next-line`

DanielMSchmidt replied:
and therefore ignore the param being missing. or if it's unused just remove the function alltogether

GeorgiSTodorov replied:
The function is useful for testing this PR and I think Mike also made use of it.

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+        entity: "package",
+        version: "v1"
+      }),
+      Accept: getContentType({
```
The `getContentType` name is strange if we get it for the `Accept` key

GeorgiSTodorov replied:
I don't understand, can you explain?

DanielMSchmidt commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,145 @@
+    )
+    .filter(([dismissed, nextVersion]) => {
+      const newVersion = getVersionFromVersionObject(nextVersion);
+      setInLocalStorage(LAST_TIME_CHECK, new Date().getTime());
```
Having a side effect in a filter seems strange, it should rather be in a `.do` block

natmegs commented on src/js/pages/Index.js:
```diff
@@ -72,6 +73,7 @@ var Index = React.createClass({
 
     ConfigStore.fetchCCID();
     ConfigStore.addChangeListener(EventTypes.CONFIG_ERROR, this.onConfigError);
+    compare.subscribe(newVersion => console.log(newVersion));
```
Maybe we could add a TODO comment here to clean up the log once this is handled. Also, we need to assign this line to a `Subscription` object so that we can `unsubscribe` somewhere (could be in `componentWillUnmount`)

GeorgiSTodorov replied:
Good points, the `console.log` was just for testing and unsubscribing should be possible.

GeorgiSTodorov commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,158 @@
+}
+
+interface DCOSUIVersion {
+  dismissedVersion: string | number | null;
```
Because `getFromLocalStorage` need to return a string or a number, since I use it for both the version and the last time check. I added a potential string type change here - https://github.com/dcos/dcos-ui/pull/3398/files/b3ddf954ad5b7083f5b0d7a79652ad5b4632de89#diff-f53b1374c8ef3fe10aefc772e8eef6ccR146 and I couldn't break it while testing it.

GeorgiSTodorov commented:
I decided to remove the `subscribe` from `pages/Index.js`, because it was breaking integration tests for some reason and it will probably be used in a different way in the final version, judging by @mperrotti 's work in progress branch - https://github.com/dcos/dcos-ui/compare/gt/feat/DCOS-41419-check-dcos-ui-update...mp/feat/DCOS-41418-ui-upgrade-banner?expand=1 . I also updated the testing instructions to reflect this.

natmegs commented on src/js/core/UpdateStream.ts:
```diff
@@ -0,0 +1,158 @@
+    .filter(([dismissed, nextVersion]) => {
+      const newVersion = getVersionFromVersionObject(nextVersion);
+
+      return compareVersions(newVersion, String(dismissed)) === 1;
```
If the dismissed version is `null` (which is currently allowed by our types), this errors with `Uncaught Error: Invalid argument not valid semver ('null' received)`. I think we either need to enforce that this is not null in the typing or do something like `return dismissed ? compareVersions(newVersion, String(dismissed)) === 1 : true;`

natmegs replied:
Also even if the dismissed version is a string but `not valid semver` this will error

GeorgiSTodorov replied:
You are right. When I wrote it like that I was thinking that `String(null)` is `"null"`, but I didn't consider that `comapreVersions` expect a specific string.

GeorgiSTodorov commented:
@natmegs , thank you for the feedback. I decided to address this issue with a try and catch block.
